### PR TITLE
Added more descriptive error to run(``) (Fixes #19094)

### LINKED ISSUE #19094: run(``): BoundsError
``` julia
julia> run(``)
ERROR: BoundsError: attempt to access 0-element Array{String,1} at index [1]
 in #414 at .\process.jl:478 [inlined]
 in setup_stdio(::Base.##414#415{Cmd,Ptr{Void},Base.Process}, ::Tuple{RawFD,RawFD,RawFD}) at .\process.jl:466
 in #spawn#413(::Nullable{Base.ProcessChain}, ::Function, ::Cmd, ::Tuple{RawFD,RawFD,RawFD}, ::Bool, ::Bool) at .\process.jl:477
 in run(::Cmd) at .\process.jl:591
```

### PR DESCRIPTION
**Pull request by SaschaMann**
As discussed in #19094 This PR adds a more descriptive error message when `run()` is called with an empty command:

``` julia
julia> run(``)
ERROR: ArgumentError: Empty command cannot be run
 in run(::Cmd) at ./process.jl:651
```

### CODE REVIEW

File: base/process.jl [hunk 1]
```diff
@@ -647,6 +647,9 @@ Run a command object, constructed with backticks. Throws an error if anything go
 including the process exiting with a non-zero status.
 """
 function run(cmds::AbstractCmd, args...)
+    if isempty(cmds.exec) # Issue 19094
```
Comment (StefanKarpinski) [inline base/process.jl]: I would not put the issue number in a comment here – we do that in test code but not in base code.

**Review by StefanKarpinski — CHANGES_REQUESTED** [review]

File: base/process.jl (hunk 1 above)
Comment (yuyichao) [inline base/process.jl]: The check seems wrong given that the function accept not only `Cmd` objects.

**Review by yuyichao — CHANGES_REQUESTED** [review]

File: base/process.jl (hunk 1 above)
Comment (SaschaMann) [inline base/process.jl]: Ah okay, I saw it in other files every now and then so I thought it'd be some undocumented convention. Removed it, should be explained by the message it throws

File: base/process.jl [hunk 2]
```diff
@@ -647,11 +647,15 @@ Run a command object, constructed with backticks. Throws an error if anything go
 including the process exiting with a non-zero status.
 """
 function run(cmds::AbstractCmd, args...)
-    if isempty(cmds.exec) # Issue 19094
-        throw(ArgumentError("Empty command cannot be run"))
+    try
+        ps = spawn(cmds, spawn_opts_inherit(args...)...)
+        success(ps) ? nothing : pipeline_error(ps)
+    catch exc
+        if isa(exc, BoundsError)
```
Comment (SaschaMann) [inline base/process.jl]: AFAIK This can only throw a BoundsError if cmds is empty, so this should be fine to check? Correct me if I'm wrong, Thanks!

Comment (tkelman) [inline base/process.jl]: It's possible that any of the other functions might happen to throw a BoundsError. Wouldn't it be better to check if `cmds` is empty, rather than using a try-catch?

File: test/process.jl [hunk 3]
```diff
@@ -0,0 +1,7 @@
+# This file is a part of Julia. License is MIT: http://julialang.org/license
+
+@testset "run(``) tests" begin
```
Comment (tkelman) [inline test/process.jl]: better to put these in an existing file (or, in a different PR, split an existing test file apart so it more closely matches the `base/` file names) - any new test file won't actually run unless it's added to `choosetests.jl`

File: base/process.jl (hunk 2 above)
Comment (StefanKarpinski) [inline base/process.jl]: I agree that this should not be done with try/catch. How about putting the check and error raising in the appropriate `spawn` method? That way you'll already be in the appropriate case based on the type of `cmds`.

File: test/process.jl (hunk 3 above)
Comment (SaschaMann) [inline test/process.jl]: Moved them to `spawn.jl` below the equality tests for `Cmd` and `AndCmds`.

File: base/process.jl (hunk 2 above)
Comment (SaschaMann) [inline base/process.jl]: Makes sense, I'll do that (tomorrow though, getting late here). Thanks for the help!

File: base/process.jl [hunk 4]
```diff
@@ -647,8 +647,15 @@ Run a command object, constructed with backticks. Throws an error if anything go
...
 """
 function run(cmds::AbstractCmd, args...)
-    ps = spawn(cmds, spawn_opts_inherit(args...)...)
-    success(ps) ? nothing : pipeline_error(ps)
+    try
+        ps = spawn(cmds, spawn_opts_inherit(args...)...)
+        success(ps) ? nothing : pipeline_error(ps)
+    catch exc
+        if isa(exc, BoundsError)
+            throw(ArgumentError("Empty command cannot be run"))
```
Comment (ararslan) [inline base/process.jl]: I'd rephrase this slightly as "cannot run empty command" (with lowercase "c" in "cannot").

**Comment by SaschaMann** [comment]
The other spawn methods all call `spawn(::Cmd, ...)` so adding the check to only that one should be enough, I think.

Also added some more tests and changed the phrasing of the error message.

File: base/process.jl [hunk 5]
```diff
@@ -647,15 +650,8 @@ Run a command object, constructed with backticks. Throws an error if anything go
...
-        ps = spawn(cmds, spawn_opts_inherit(args...)...)
-        success(ps) ? nothing : pipeline_error(ps)
-    catch exc
-        if isa(exc, BoundsError)
-            throw(ArgumentError("Empty command cannot be run"))
-        end
-        rethrow()
-    end
+    ps = spawn(cmds, spawn_opts_inherit(args...)...)
+    success(ps) ? nothing : pipeline_error(ps)
```
Comment (SaschaMann) [inline base/process.jl]: This just restores it to what it was originally.

**Review by tkelman — APPROVED** [review]
lgtm (should be squashed on merge)

**Comment by ararslan** [comment]
Looks good to me--great job here! @StefanKarpinski and @yuyichao, would you two mind taking another look, since your requests appear to have been addressed?

**Review by yuyichao — APPROVED** [review]

**Review by StefanKarpinski — APPROVED** [review]
